src/pybind/mgr/cephadm/module.py

```diff
@@ -1984,21 +1984,15 @@ def _create_daemon(self, daemon_type, daemon_id, host,
                        keyring=None,
                        extra_args=None, extra_config=None,
                        reconfig=False,
-                       osd_uuid_map=None):
+                       osd_uuid_map=None,
+                       custom_config=None,  # type: Optional[Callable[[], Dict[str, Any]]]
+                       ):
         if not extra_args:
             extra_args = []
         name = '%s.%s' % (daemon_type, daemon_id)
 
-        if daemon_type == 'prometheus':
-            j = self._generate_prometheus_config()
-            extra_args.extend(['--config-json', '-'])
-        elif daemon_type == 'node-exporter':
-            j = None
-        elif daemon_type == 'grafana':
-            j = self._generate_grafana_config()
-            extra_args.extend(['--config-json', '-'])
-        elif daemon_type == 'alertmanager':
-            j = self._generate_alertmanager_config()
+        if callable(custom_config):
+            cephadm_config = custom_config()
```

Comment:
maybe instead
```
if daemon_type in ['prometheus', 'node-exporter', 'grafana', 'alertmanager']:
    if custom_config:
        cephadm_config = custom_config()
        extra_args.extend(['--config-json', '-'])
    else:
        cephadm_config = ''
```
or something?  that seems less awkawrd than passing a no-op config function for node-exporter

Reply:
I've had something like that in mind, but forgot to implement it. As `custom_config` is allowed to be `None`, I can simply remove the empty function parameter for node-exporter. That's what I'd prefer as one point of this refactoring I had in mind is removing the daemon specific code from `_create_daemon`. What do you think?

Reply:
hmm actually now that i look more closely i think it needs to stay in _create_daemon because of these two callers:
```
    @async_map_completion
    def _daemon_action(self, daemon_type, daemon_id, host, action):
        if action == 'redeploy':
            # stop, recreate the container+unit, then restart
            return self._create_daemon(daemon_type, daemon_id, host)
        elif action == 'reconfig':
            return self._create_daemon(daemon_type, daemon_id, host,
                                       reconfig=True)
```

Reply:
...which underscores the point that the config should be drawn from config-key and not extra parameters in the orch CLI.